DrillSideways optimizations

gsmiller opened the pull request:
### Description

This change makes use of `advance` instead of `next` where possible and splits out 1st and 2nd phase checking to avoid match confirmation when unnecessary.

Note that I only focused on the `doQueryFirstScoring` implementation here and didn't modify the other two scoring approaches. "Progress not perfection" and all that (plus, I think we should strongly consider removing these other two implementations, but we'd want to benchmark to be certain).

Unfortunately, `luceneutil` doesn't have dedicated drill sideways benchmarks, but some benchmarks on our internal software that makes use of drill sideways showed a +2% QPS improvement and no obvious regressions.

gsmiller committed:
DrillSideways uses advance instead of next when multiple dims miss

gsmiller committed:
changes

gsmiller committed:
more optimizations

gsmiller committed:
changes update and spotless

zhaih commented on lucene/facet/src/java/org/apache/lucene/facet/DrillSidewaysScorer.java:
```diff
@@ -149,63 +162,147 @@ public int score(LeafCollector collector, Bits acceptDocs, int min, int maxDoc)
+            assert dim.approximation.docID() == docID;
+            if (dim.twoPhase.matches() == false) {
+              if (failedDim != null) {
+                assert dim.approximation.docID() + 1 <= failedDim.approximation.docID();
```
Why is this assertion true? Shouldn't the two docID be the same and both equal to `docID` according to assertion in L228?

gsmiller replied:
Great catch, thanks! I think this must be a silly copy/paste on my part. I can't image what I was thinking otherwise. Also, we can be more aggressive with setting the next docID here. I'll update the PR with these changes. I've actually been working on another change that builds on this (on a separate fork) which allows sideways facets collectors to terminate early and then fold their conditions into hard requirements on the base query. That's a whole separate change to propose another day, but what slipped through here is an unfortunate side-effect of trying to keep two versions of a partial change in sync. Anyway, thanks again for the catch!

zhaih commented on lucene/facet/src/java/org/apache/lucene/facet/DrillSidewaysScorer.java:
```diff
@@ -149,63 +162,147 @@ public int score(LeafCollector collector, Bits acceptDocs, int min, int maxDoc)
+      if (twoPhaseDims != null) {
+        if (failedDim == null) {
+          for (DocsAndCost dim : twoPhaseDims) {
+            assert dim.approximation.docID() == docID;
```
I found it easier to understand if we assert `dim.approximation.docID() == baseApproximation.docID()` or simply add another line asserting `docID == baseApproximation.docID()`?

gsmiller replied:
Sure, thanks!

zhaih reviewed (commented):
Overall LGTM except for one assertion, also do we have existing unit test testing those code? (I hope there's code coverage report here :)

gsmiller committed:
pr feedback and slightly more aggressive doc advance in two-phase misses

zhaih commented:
Changes LGTM, do we need to add some unit tests?

gsmiller commented:
> Changes LGTM, do we need to add some unit tests?

Thanks @zhaih. Let me consider some specific test for this. I know our randomized testing for DrillSideways covers these code paths but maybe some specific, non-random tests would be useful.

gsmiller committed:
static comparators

gsmiller commented:
@zhaih I reexamined our test coverage and think we're in good shape already actually. We've got good coverage for covering drill-sideways correctness with multiple dimensions, etc. (including random and non-random). We could try to take these further by somehow asserting that advance is being used in favor of nextDoc when appropriate, but I think those tests would be reasonably complex to write and I'm not sure they add tremendous value. I'd rather we spent time building drill-sideways benchmarks that focus on ensuring our performance doesn't regress. But that's just my opinion. Please let me know if you feel differently and we can keep discussing. Thanks!

zhaih reviewed (approved):
Hi @gsmiller, I don't have strong opinion on adding a test because I don't really know what coverage we already got, one fact that makes me a little worry about the coverage is that assertion caught by me should've been caught by unit test and that's why I'm asking. But if you think that'll be covered by randomized test then please feel free to push it!

gsmiller commented:
@zhaih that's a good point and valid concern. I dug into the existing tests and it looks like we have lots of coverage _except_ that the majority of the coverage is using basic, single-phase drill-down dimensions. I'm going to augment our randomized testing to randomly use two-phase drill-downs to broaden coverage. Thanks for the discussion!

gsmiller committed:
1. augment randomized testing to cover two-phse case
2. remove the "validateState" assertion since it's illegal to call match() more than one for the same doc (state validation would require separately tracking the match results for all two-phase iterators, which doesn't seem worth it)

zhaih commented:
@gsmiller Thank you for checking and continuous effort!

gsmiller commented:
@zhaih well, thank you for keeping me honest with testing. I think I've already found an insidious, potential bug with some beefier tests.

zhaih reviewed (approved):
Thank you, test LGTM!
